src/dataloader/mod.rs

```diff
@@ -0,0 +1,614 @@
+mod cache;
+
+use std::any::TypeId;
+use std::borrow::Cow;
+use std::collections::{HashMap, HashSet};
+use std::hash::Hash;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+pub use cache::{CacheFactory, CacheStorage, HashMapCache, LruCache, NoCache};
+use fnv::FnvHashMap;
+use futures_channel::oneshot;
+use futures_timer::Delay;
+use futures_util::future::BoxFuture;
+#[cfg(feature = "tracing")]
+use tracing::{info_span, instrument, Instrument};
+#[cfg(feature = "tracing")]
+use tracinglib as tracing;
+
+#[allow(clippy::type_complexity)]
+struct ResSender<K: Send + Sync + Hash + Eq + Clone + 'static, T: Loader<K>> {
+  use_cache_values: HashMap<K, T::Value>,
+  tx: oneshot::Sender<Result<HashMap<K, T::Value>, T::Error>>,
+}
+
+struct Requests<K: Send + Sync + Hash + Eq + Clone + 'static, T: Loader<K>> {
+  keys: HashSet<K>,
+  pending: Vec<(HashSet<K>, ResSender<K, T>)>,
+  cache_storage: Box<dyn CacheStorage<Key = K, Value = T::Value>>,
+  disable_cache: bool,
+}
+
+type KeysAndSender<K, T> = (HashSet<K>, Vec<(HashSet<K>, ResSender<K, T>)>);
+
+impl<K: Send + Sync + Hash + Eq + Clone + 'static, T: Loader<K>> Requests<K, T> {
+  fn new<C: CacheFactory>(cache_factory: &C) -> Self {
+    Self {
+      keys: Default::default(),
+      pending: Vec::new(),
+      cache_storage: cache_factory.create::<K, T::Value>(),
+      disable_cache: false,
+    }
+  }
+
+  fn take(&mut self) -> KeysAndSender<K, T> {
+    (std::mem::take(&mut self.keys), std::mem::take(&mut self.pending))
+  }
+}
+
+/// Trait for batch loading.
+#[async_trait::async_trait]
+pub trait Loader<K: Send + Sync + Hash + Eq + Clone + 'static>: Send + Sync + 'static {
+  /// type of value.
+  type Value: Send + Sync + Clone + 'static;
+
+  /// Type of error.
+  type Error: Send + Clone + 'static;
+
+  /// Load the data set specified by the `keys`.
+  async fn load(&self, keys: &[K]) -> Result<HashMap<K, Self::Value>, Self::Error>;
+}
+
+struct DataLoaderInner<K: Send + Sync + Hash + Eq + Clone + 'static, T: Loader<K>> {
+  requests: Mutex<FnvHashMap<TypeId, Requests<K, T>>>,
```

Comment:
will anyway leave a comment here just as a reminder

in next iteration we should try to remove wrapper `FnvHashMap` over Requests since there is not possible anymore to have different types for the same loader

Reply:
made this change and it's working now